File: openbis_api/source/java/ch/ethz/sis/openbis/generic/shared/api/v3/dto/search/DataSetSearchCriterion.java
/*
 * Copyright 2014 ETH Zuerich, CISD
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.ethz.sis.openbis.generic.shared.api.v3.dto.search;

import ch.ethz.sis.openbis.generic.shared.api.v3.dto.id.dataset.IDataSetId;
import ch.systemsx.cisd.base.annotation.JsonObject;

/**
 * @author pkupczyk
 */
@JsonObject("DataSetSearchCriterion")
public class DataSetSearchCriterion extends AbstractEntitySearchCriterion<IDataSetId>
{

    private static final long serialVersionUID = 1L;

    private DataSetSearchRelation relation;

    public DataSetSearchCriterion()
    {
        this(DataSetSearchRelation.DATASET);
    }

    DataSetSearchCriterion(DataSetSearchRelation relation)
    {
        this.relation = relation;
    }

    public DataSetSearchCriterion withParents()
    {
        return with(new DataSetParentsSearchCriterion());
    }

    public DataSetSearchCriterion withChildren()
    {
        return with(new DataSetChildrenSearchCriterion());
    }

    public DataSetSearchCriterion withContainer()
    {
        return with(new DataSetContainerSearchCriterion());
    }

    public DataSetSearchCriterion withOrOperator()
    {
        return (DataSetSearchCriterion) withOperator(SearchOperator.OR);
    }

    public DataSetSearchCriterion withAndOperator()
    {
        return (DataSetSearchCriterion) withOperator(SearchOperator.AND);
    }

    public DataSetSearchRelation getRelation()
    {
        return relation;
    }

    @Override
    protected SearchCriterionToStringBuilder createBuilder()
    {
        SearchCriterionToStringBuilder builder = super.createBuilder();
        builder.setName(relation.name());
        return builder;
    }

}

File: js-test/servers/common/core-plugins/tests/1/as/webapps/openbis-v3-api-test/html/models/dto/search/DataSetSearchCriterion.js
/**
 * @author pkupczyk
 */
define([ "support/stjs", "dto/search/AbstractEntitySearchCriterion", "dto/search/DataSetSearchRelation", "dto/search/ExperimentSearchCriterion", "dto/search/SampleSearchCriterion",
		"dto/search/SearchOperator" ], function(stjs, AbstractEntitySearchCriterion, DataSetSearchRelation, ExperimentSearchCriterion, SampleSearchCriterion, SearchOperator) {
	var DataSetSearchCriterion = function(relation) {
		AbstractEntitySearchCriterion.call(this);
		this.relation = relation ? relation : DataSetSearchRelation.DATASET;
	};
	stjs.extend(DataSetSearchCriterion, AbstractEntitySearchCriterion, [ AbstractEntitySearchCriterion ], function(constructor, prototype) {
		prototype['@type'] = 'dto.search.DataSetSearchCriterion';
		constructor.serialVersionUID = 1;
		prototype.relation = null;
		prototype.withParents = function() {
			return this.addCriterion(new DataSetParentsSearchCriterion());
		};
		prototype.withChildren = function() {
			return this.addCriterion(new DataSetChildrenSearchCriterion());
		};
		prototype.withContainer = function() {
			return this.addCriterion(new DataSetContainerSearchCriterion());
		};
		prototype.withExperiment = function() {
			return this.addCriterion(new ExperimentSearchCriterion());
		};
		prototype.withSample = function() {
			return this.addCriterion(new SampleSearchCriterion());
		};
		prototype.withOrOperator = function() {
			return this.withOperator(SearchOperator.OR);
		};
		prototype.withAndOperator = function() {
			return this.withOperator(SearchOperator.AND);
		};
		prototype.getRelation = function() {
			return this.relation;
		};
		prototype.createBuilder = function() {
			var builder = AbstractEntitySearchCriterion.prototype.createBuilder.call(this);
			builder.setName(this.relation.name());
			return builder;
		};
	}, {
		relation : {
			name : "Enum",
			arguments : [ "DataSetSearchRelation" ]
		},
		operator : {
			name : "Enum",
			arguments : [ "SearchOperator" ]
		},
		criteria : {
			name : "Collection",
			arguments : [ "ISearchCriterion" ]
		}
	});

	var DataSetParentsSearchCriterion = function() {
		DataSetSearchCriterion.call(this, DataSetSearchRelation.PARENTS);
	};
	stjs.extend(DataSetParentsSearchCriterion, DataSetSearchCriterion, [ DataSetSearchCriterion ], function(constructor, prototype) {
		prototype['@type'] = 'dto.search.DataSetParentsSearchCriterion';
		constructor.serialVersionUID = 1;
	}, {
		relation : {
			name : "Enum",
			arguments : [ "DataSetSearchRelation" ]
		},
		operator : {
			name : "Enum",
			arguments : [ "SearchOperator" ]
		},
		criteria : {
			name : "Collection",
			arguments : [ "ISearchCriterion" ]
		}
	});

	var DataSetChildrenSearchCriterion = function() {
		DataSetSearchCriterion.call(this, DataSetSearchRelation.CHILDREN);
	};
	stjs.extend(DataSetChildrenSearchCriterion, DataSetSearchCriterion, [ DataSetSearchCriterion ], function(constructor, prototype) {
		prototype['@type'] = 'dto.search.DataSetChildrenSearchCriterion';
		constructor.serialVersionUID = 1;
	}, {
		relation : {
			name : "Enum",
			arguments : [ "DataSetSearchRelation" ]
		},
		operator : {
			name : "Enum",
			arguments : [ "SearchOperator" ]
		},
		criteria : {
			name : "Collection",
			arguments : [ "ISearchCriterion" ]
		}
	});

	var DataSetContainerSearchCriterion = function() {
		DataSetSearchCriterion.call(this, DataSetSearchRelation.CONTAINER);
	};
	stjs.extend(DataSetContainerSearchCriterion, DataSetSearchCriterion, [ DataSetSearchCriterion ], function(constructor, prototype) {
		prototype['@type'] = 'dto.search.DataSetContainerSearchCriterion';
		constructor.serialVersionUID = 1;
	}, {
		relation : {
			name : "Enum",
			arguments : [ "DataSetSearchRelation" ]
		},
		operator : {
			name : "Enum",
			arguments : [ "SearchOperator" ]
		},
		criteria : {
			name : "Collection",
			arguments : [ "ISearchCriterion" ]
		}
	});

	return DataSetSearchCriterion;
})